lib/puppet/type/mount.rb

```diff
@@ -293,5 +293,8 @@ def value(name)
       dependencies[0..-2]
     end
 
+    # Autorequire the mount point's file resource
+    autorequire(:file) { Pathname.new(@parameters[:name].value) }
```

Comment:
I'm +1 to moving the logic to mount, In addition to what @MikaelSmith said, I think it's worth doing in mount just from a code-cleanliness standpoint. There are actually two sides to this:

* Mount should depend on the file resource that creates the mountpoint directory
* Mount should happen before any files on that mountpoint

Keeping all that in one place seems cleaner to me than having it split between file and mount.